src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1162 @@
+    memset(d_completed_md5, TOKEN_STATE_PENDING, sizeof(d_completed_md5));
+
+    d_total_ingressed_obj = 0;
+    d_num_failed_workers = 0;
+  }
+
+
+  static constexpr auto COOKIE_LEN = 15;
+  static constexpr auto CLUSTER_ID_LEN = 15;
+  //---------------------------------------------------------------------------
+  cluster::cluster(const DoutPrefixProvider *_dpp,
+                   CephContext *cct,
+                   rgw::sal::Driver* driver):
+    dpp(_dpp),
+    d_lock_cookie(gen_rand_alphanumeric(cct, COOKIE_LEN)),
+    d_cluster_id (gen_rand_alphanumeric(cct, CLUSTER_ID_LEN))
+  {
+    clear();
+    ldpp_dout(dpp, 10) << __func__ << "::cluser_id=" << d_cluster_id << dendl;
+
+    auto store = dynamic_cast<rgw::sal::RadosStore*>(driver);
+    if (!store) {
+      ldpp_dout(dpp, 0) << "ERR: failed dynamic_cast to RadosStore" << dendl;
+      ceph_abort("non-rados backend");
+      return;
+    }
+
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      throw std::runtime_error("Failed init_dedup_pool_ioctx()");
+    }
+
+    // generate an empty epoch with zero counters
+    int ret = set_epoch(ioctx, d_cluster_id, dpp, 0, 0);
+    if (ret != 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::failed set_epoch()! ret="
+                        << ret << "::" << cpp_strerror(-ret) << dendl;
+      throw std::runtime_error("Failed set_epoch()");
+    }
+
+    dedup_epoch_t epoch;
+    reset(store, ioctx, &epoch, 0, 0);
+  }
+
+  //---------------------------------------------------------------------------
+  int cluster::reset(rgw::sal::RadosStore *store,
+                     librados::IoCtx &ioctx,
+                     dedup_epoch_t *p_epoch,
+                     work_shard_t num_work_shards,
+                     md5_shard_t num_md5_shards)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::REQ num_work_shards=" << num_work_shards
+                       << "::num_md5_shards=" << num_md5_shards << dendl;
+    clear();
+
+    while (true) {
+      int ret = get_epoch(ioctx, dpp, p_epoch, __func__);
+      if (ret != 0) {
+        return ret;
+      }
+      if (p_epoch->num_work_shards && p_epoch->num_md5_shards) {
+        ldpp_dout(dpp, 10) << __func__ << "::ACC num_work_shards=" << p_epoch->num_work_shards
+                           << "::num_md5_shards=" << p_epoch->num_md5_shards << dendl;
+        break;
+      }
+      else if (!num_work_shards && !num_md5_shards) {
+        ldpp_dout(dpp, 10) << __func__ << "::Init flow, no need to wait" << dendl;
+        break;
+      }
+      else {
+        ret = swap_epoch(dpp, ioctx, p_epoch,
+                         static_cast<dedup_req_type_t> (p_epoch->dedup_type),
+                         num_work_shards, num_md5_shards);
+      }
+    }
+
+    d_epoch_time = p_epoch->time;
+    cleanup_prev_run(ioctx);
+    create_shard_tokens(ioctx, p_epoch->num_work_shards, WORKER_SHARD_PREFIX);
```

Comment:
these functions are always returning success (0).
don't we want to check if token creation failed and propagate it here?

Reply:
This code is running in parallel by multiple independent RGW so failures are expected (e.g. 2 RGW members trying to remove the same object)

Maybe I should add a step to verify that all old objects were removed and that all new tokens were created

Reply:
`create_shard_tokens()` already takes care of parallel execution by checking for `-EEXIST` explicitly.
however, if the token object is not created, will the dedup be able to run?

Reply:
We can't run without all tokens created.
I had transient errors caused by parallel running, I don't recall the details.
I think the safest approach is to run a verify loop that all needed tokens exist and that their creation time is after EPOCH (which means delete was successful)

Reply:
the transient errors are already handled by checking for `-EEXIST`.

but the verify loop approach is probabnly better (e.g. one RGW failed to create the token due to some issue, but another was successfull).

Reply:
done